---
title: "flat_first.Rmd for working package"
output: html_document
editor_options: 
  chunk_output_type: console
---

<!-- Run this 'development' chunk -->
<!-- Store every call to library() that you need to explore your functions -->

```{r development, include=FALSE}
library(testthat)
```

<!--
 You need to run the 'description' chunk in the '0-dev_history.Rmd' file before continuing your code there.

If it is the first time you use {fusen}, after 'description', you can directly run the last chunk of the present file with inflate() inside.
--> 

```{r development-load}
# Load already included functions if relevant
pkgload::load_all(export_all = FALSE)
```

# Include some data examples in your package

<!-- 
 Store your dataset in a directory named "inst/" at the root of your project.
 Use it for your tests in this Rmd thanks to `pkgload::load_all()` to make it available
and `system.file()` to read it in your examples.

- There already is a dataset in the "inst/" directory to be used in the examples below
-->

```{r development-dataset}
# Run all this chunk in the console directly
# There already is a dataset in the "inst/" directory
# Make the dataset file available to the current Rmd during development
pkgload::load_all(path = here::here(), export_all = FALSE)

# You will be able to read your example data file in each of your function examples and tests as follows - see chunks below
datafile <- system.file("BesePopTest.rds", package = "RITHMS")
ToyData <- readRDS(datafile)
```

# Pre-Simulation functions 

## Formatting data functions

```{r function-formatting_data}
#' Formatting data from file paths to base population object
#'
#' @param path_to_microbiome String giving the path to count table file
#' @param path_to_pedmap String giving the path and prefix to ped and map file
#' @param threshold Threshold for rarefaction, DEFAULT = 0.05
#' @param ind_selected Vector of string values with individuals to keep, have to match rownames of count table file, DEFAULT = NULL
#'
#' @return
#' A data.frame corresponding to (rarefied) microbiome with individuals in rows and taxa in columns. Genotypes data.frame is an attribute called "population" reachable using attr(output_name,"population")
#' @export
read_input_data <- function(path_to_microbiome,
                            path_to_pedmap,
                            threshold = 0.05,
                            ind_selected = NULL){
  ########
  #read microbiome file and make genotypes and microbiome data match according to individuals
  ########
  microbiome <- fread(path_to_microbiome)
  microbiome_filtered5 <- rarefied_microbiome(microbiome, threshold = threshold)
  if(!is.null(ind_selected)){
    microbiome_filtered5 <- microbiome_filtered5 %>% filter(rownames(.) %in% ind_selected)
  }
  founder_object <- generate_founder(path = glue("{path_to_pedmap}"),
                                     microbiome_filtered5)
  return(founder_object)
}


```

```{r function-useful_formatting_data}
#' Rarefaction step within read_input_data
#'
#' @inheritParams read_input_data
rarefied_microbiome <- function(microbiome_matrix, 
                                threshold = 0.05){
  microbiome_matrix <- microbiome_matrix |> as_tibble() |> tibble::column_to_rownames("numero_individu")
  abundant_otus <- tibble(
    V1         = colnames(microbiome_matrix),
    prevalence = colMeans(microbiome_matrix > 0)
  ) |> rename(OTU = V1) |> filter(prevalence >= threshold) |> pull(OTU)
  return(microbiome_matrix[, abundant_otus])
}

#' Formatting of ped and map files into haplotypes table and filter individuals
#'
#' @inheritParams read_input_data
generate_founder <- function(path = NULL,
                             microbiote_matrix){
  # # Convert PED-file into haplotype dataset (one haplotype per colum)
  if(is.null(path)){
    stop("path must be a valid string like : '/path/to/directory/prefix'")
  }else{
    if(FALSE %in% file_test("-f", c(glue::glue("{path}.ped"),
                                    glue::glue("{path}.map")))){
      stop(glue::glue("Files {path}.ped and/or {path}.map don't exist, please check your path argument"))
    }else{
      map <- as.matrix(read.table(glue::glue("{path}.map")))
      ped <- as.matrix(read.table(glue::glue("{path}.ped")))
      #Step to filter ped map based on microbiome matrix,
      #rownames microbiote_matrix must be ind ID to match column 2 of ped file
      ped_ind <- ped[,2] %>% gsub("\\s","",.)
      microbiote_ind <- rownames(microbiote_matrix)
      microbiote_filtered <- microbiote_matrix[microbiote_ind %in% ped_ind,]
      ped <- ped[ped_ind %in% microbiote_ind,]
      ####
      nsnp <- (ncol(ped)-6)/2
      haplo1 <- ped[,1:nsnp*2+6-1] #all odd numbers
      haplo2 <- ped[,1:nsnp*2+6] #all uneven numbers
      haplo <- t(rbind(haplo1, haplo2)[c(0,nrow(haplo1)) + sort(rep(1:nrow(haplo1),2)),])
      #haplo <- ifelse(haplo == "0", NA,haplo)
      population <- creating.diploid(dataset = haplo, map = map, verbose=verbose)
      attr(microbiote_filtered,"population") <- population
      return(microbiote_filtered)
    }
  }
}

```

## Taxa assignation function

```{r function-taxa_assignment}
#' Assign all taxa to a cluster, eventually under genetic control using hclust
#'
#' @param founder_object Output of read_input_data function
#' @param taxa_g Percentage of taxa under genetic control, DEFAULT = 0.1
#' @inheritParams stats::hclust
#' @inheritParams stats::cutree
#' @inheritParams vegan::vegdist
#'
#' @return
#' A vector with a length matching the total number of taxa with values from 0 to the number of taxa, 0 corresponding to the non under genetic control cluster
#' @rdname formatting_data
#' @export

assign_taxa <- function(founder_object,
                        taxa_g = 0.05){
  dist_mat <- vegan::vegdist(x = t(founder_object),
               method = "bray")
  hc <- hclust(dist_mat)
  taxa_assign <- cutree(hc, k = 100)
  ntaxa_g = 0
  cluster_id_g = c()
  if(sum(table(taxa_assign)[table(taxa_assign) > 10 & table(taxa_assign) < 25]) < round(length(taxa_assign)*taxa_g)){
    print(glue("{sum(table(taxa_assign)[table(taxa_assign) > 10 & table(taxa_assign) < 25])} taxa available for genetic control sampling but {round(length(taxa_assign)*taxa_g)} asked. Change taxa_g parameter please."))
  }
  while(ntaxa_g < round(length(taxa_assign)*taxa_g)){
    new_id <- table(taxa_assign)[table(taxa_assign) > 10 & table(taxa_assign) < 25] %>% sample(1) %>% names() %>% as.numeric()
    if(! new_id %in% cluster_id_g){
      cluster_id_g <- append(cluster_id_g, new_id)
      ntaxa_g <- ntaxa_g + sum(taxa_assign == new_id)
    }
  }
  taxa_assign_g <- ifelse(taxa_assign %in% cluster_id_g, taxa_assign, 0) %>% as.factor()
  levels(taxa_assign_g) <- c(0:length(cluster_id_g))
  return(taxa_assign_g)
}

```

```{r examples-taxa_assignment}
# library(dplyr)
# datafile <- system.file("BesePopTest.rds", package = "RITHMS")
# ToyData <- readRDS(datafile)
# assign_taxa(founder_object = ToyData,
#             taxa_g = 0.2)
```

```{r tests-taxa_assignment}
#test length of assignment vector
```

## Calibration of genetic effect

```{r function-genetic_effect}
#' Calibration of genetic effect from founder population data 
#'
#' @param founder_object Output of generate_founder() function
#' @param taxa_assign_g Factor vector giving cluster assignment for all taxa, typical output of assign_taxa()
#' @param correlation Correlation between taxa within the same cluster, value between 0 and 1, DEFAULT = 0.5
#' @param effect.size Vector giving the size of genetic effect to try 
#' @param plot boolean, if plot generation is required, DEFAULT = TRUE
#'
#' @return
#' A data.frame with three columns, giving the Taxa ID, the effect.size and the corresponding heritability
#' @export
gen_effect_calibration <- function(founder_object,
                                   taxa_assign_g,
                                   correlation = 0.5,
                                   effect.size = c(seq(0.1,1,0.1)),
                                   plot = T){
  #init of all parameters
  
  #Microbiome part
  M <- founder_object %>% t() %>% apply(2, \(x) x/sum(x))
  M0_clr <- M |> t() |> compositions::clr() |> t()
  taxa_scale <- apply(M0_clr, 1, sd)
  #Genotype
  G <- get.geno(attr(founder_object,"population"),gen=1)
  
  #beta arguments
  n_g <- nrow(G)
  n_clust_g <- taxa_assign_g[taxa_assign_g != 0] %>% unique() %>% length()
  n_qtl_o <- ((n_g * 0.2)/n_clust_g)  %>% round()
  
  #output variables 
  out_data <- data.frame()
  
  #One run to obtain beta info matrix 
  beta_innit <- compute_beta_matrix_cluster(n_b = nrow(M), 
                                            n_g = n_g, 
                                            n_clust = taxa_assign_g, 
                                            n_qtl_o = n_qtl_o, 
                                            n_otus = sum(taxa_assign_g != 0), 
                                            correlation = correlation, 
                                            effect_size = 0.1)
  
  beta_info <- attr(beta_innit, "sim_params")
  
  for(e in effect.size){
    beta <- compute_beta_matrix_cluster(n_b = nrow(M), 
                                        n_g = n_g, 
                                        n_clust = taxa_assign_g, 
                                        n_qtl_o = n_qtl_o, 
                                        n_otus = sum(taxa_assign_g != 0), 
                                        correlation = correlation, 
                                        effect_size = e,
                                        beta_info = beta_info)
    beta_test <- attr(beta, "sim_params")
    #print(beta_test$id_otu[1])
    betaG_e <- compute_beta_g(beta, G, noise = 0.5, taxa_scale = taxa_scale)
    betaG <- attr(betaG_e,"noise")$betag
    betag_var <- betaG |> apply(1,var)
    
    M_sim <- M0_clr + betaG_e
    M_var <- apply(M_sim,1,var)
    
    out_data <- rbind(out_data,cbind(Taxa = names(M_var[betag_var != 0]), effect.size = e, Heritability = (betag_var[betag_var != 0] / M_var[betag_var != 0])))
  }
  
  out_data$effect.size <- as.numeric(out_data$effect.size)
  out_data$Heritability <- as.numeric(out_data$Heritability)
  
  if(plot){
    p <- out_data %>% ggplot(aes(x = as.factor(effect.size),y=Heritability, fill = as.factor(effect.size))) +
      geom_boxplot()+
      labs(title = glue("Heritabilities distribution"),
           subtitle = glue("({length(unique(out_data$Taxa))} taxa under genetic control)"),
           x = "Genetic effect size") +
      theme(panel.background = element_rect(fill="white"),
            panel.grid.major = element_line(colour="#e3e3e3"),
            panel.grid.minor = element_line(colour="#e9e9e9"),
            axis.title = element_text(size = 12),
            axis.text = element_text(size=9),
            plot.title = element_text(size=12),
            aspect.ratio = 1.5,
            legend.position = "none")
    print(p)
    
    p2 <- out_data %>% ggplot(aes(x=Heritability,y=as.factor(effect.size),fill=as.factor(effect.size))) +
      geom_density_ridges(alpha=0.8)+
      labs(title = "Ridges plot of taxa heritability",
           x = "Taxa heritability",
           y = "Genetic effect size")+
      theme(legend.position = "none",
            panel.background = element_rect(fill="white"),
            panel.grid.major = element_line(colour="#e3e3e3"),
            panel.grid.minor = element_line(colour="#e9e9e9"),
            axis.title = element_text(size = 12),
            axis.text = element_text(size=9),
            plot.title = element_text(size=12),
            aspect.ratio = 1.5)
    print(p2)
  }
  return(out_data)
}
```

```{r examples-genetic_effect}
```

```{r tests-genetic_effect}
```

```{r function-useful_genetic_effect}
#' Generate beta matrix giving genetic effect per SNP on taxa abundances.
#'
#' @inheritParams gen_effect_calibration
compute_beta_matrix_cluster <- function(n_b,
                                        n_g,
                                        n_clust,
                                        n_qtl_o,
                                        n_otus, #percentage
                                        effect_size = 1,
                                        correlation = 1, ## value between 0 and 1
                                        beta_info = NULL
){
  # Initialisation of beta matrix of genetic effects
  beta <- matrix(0, nrow = n_b, ncol = n_g, 
                 dimnames = list(
                   glue('OTU_{1:n_b}'), 
                   glue('SNP_{1:n_g}') 
                 ))
  
  nb_k <- (n_clust[n_clust != 0] %>% unique() %>% length())  
  
  if(is.null(beta_info)){
    beta_info <- tibble(
      cluster     = 1:nb_k,
      id_otu      = split(1:length(n_clust), n_clust)[-1],
      id_qtl_o    = map(cluster, \(x) {sample(1:n_g, n_qtl_o)})
    )
  }
  sd_factor <- effect_size / sqrt(n_qtl_o)
  for(clus in 1:nb_k){
    # Extract causal SNPs for current cluster
    id_qtl_o <- beta_info[["id_qtl_o"]][clus][[1]]
    # Extract OTUs under genetic control for current cluster
    ID_OTUs_gen <- beta_info[["id_otu"]][clus][[1]]
    # fill beta with random variable 
    # TODO: should the effect of a SNP be shared across OTU in a cluster ?
    # to create blockwise effect in beta G ? 
    lambda <- root(correlation)
    ind_coef         <- rnorm(length(ID_OTUs_gen) * n_qtl_o, mean = 0, sd = sd_factor)
    correlated_coef  <- matrix(rnorm(n_qtl_o, 0, sd_factor), 
                               nrow = length(ID_OTUs_gen), ncol = n_qtl_o,
                               byrow = TRUE)
    beta[ID_OTUs_gen,id_qtl_o] <- (lambda * correlated_coef + (1 - lambda) * ind_coef) / sqrt(lambda^2 + (1-lambda)^2)
  }
  
  attr(beta, "sim_params") <- beta_info
  return(beta)
}

#' Compute product of matrixes based on few parameters
#'
#' @inheritParams gen_effect_calibration
compute_beta_g <- function(beta,
                           genotypes,
                           noise,
                           taxa_scale){
  G <- genotypes
  beta_g <- beta %*% G
  beta_g <- beta_g %>% t() %>% scale(center=TRUE,scale=FALSE) %>% t() 
  beta_g_raw <- beta_g
  noise_vec = rnorm(n = nrow(beta) * ncol(G), sd = noise) * taxa_scale
  beta_g <-  beta_g_raw + noise_vec
  attr(beta_g,"noise") <- list('noise' = noise_vec,'betag' = beta_g_raw)
  return(beta_g)
}

#' Compute root value for beta matrix construction
#'
#' @inheritParams gen_effect_calibration
root <- function(rho) {
  if (rho == 0.5) return(0.5)
  (2 * rho - sqrt(4 * rho^2 - 4 * rho * (2* rho - 1))) / (4*rho - 2)
}
``` 

# HoloSimu
## Useful functions 

```{r function-useful_holo_simu}
#' Calibration of coefficients to compute phenotypes
#'
#' @inheritParams holo_simu
calibrate_params_phenotypes <- function(X0,
                                        B0,
                                        h2,
                                        b2,
                                        otu_list,
                                        Nqtl_y,
                                        Notu_y = length(otu_list)){
  if (Notu_y == 0) b2 = 0
  if (Nqtl_y == 0) h2 = 0
  
  ## heritability and microbiability should be non-negative and the sum should be positive
  stopifnot(h2 >= 0, b2 >= 0, h2 + b2 > 0)
  
  #init 
  Nind = ncol(X0)
  Nsnp = nrow(X0)
  Notu = nrow(B0)
  Bclust=seq(Notu)
  Nclust = max(Bclust)
  gb = rep(0,Nind)
  gq = gb
  
  ## list of snps directly influencing y :
  qtl_list  = sample(seq(Nsnp), size = Nqtl_y)
  beta_qtl = rgamma(Nqtl_y, shape = 0.4, scale = 5) * sample(c(1,-1), size=Nqtl_y, replace = T)
  beta_otu = rgamma(Notu_y, shape = 1.4, scale = 3.8) * sample(c(1,-1), size = Notu_y, replace = T)
  
  # OTU part  
  if(Notu_y>0){ #arrow from B to y
    B <- as.matrix(B0)
    gb = t(B[otu_list, , drop = FALSE]) %*% beta_otu
    varb <- var(gb)
  }
  
  ## If not genetic effect:
  ## - no scaling for microbiote parameters
  ## - scale noise variance using microbiote as a reference
  if (h2 == 0) {
    
    ## genetic effects
    beta_qtl[] <- 0
    gq <- rep(0, Nind)
    
    ## noise scaling
    se <- sqrt(varb * (1-b2) / b2)
  } else {
    ## Scale microbiote and noise variance using genetic effects as reference
    
    ## genetic effects
    stopifnot(Nqtl_y > 0) ## Nqtl_y should be positive if h2 > 0
    X <- as.matrix(X0)
    gq = t(X[qtl_list, , drop = FALSE]) %*% beta_qtl
    varg <- var(gq)
    
    ## microbiote and noise scaling
    microbiote_scaling <- sqrt( (b2 * varg) / (h2 * varb) )
    beta_otu[] <- microbiote_scaling * beta_otu
    se <- sqrt(varg * (1- h2 - b2) / h2)
  }
  
  ## Rescale all coefficents to ensure that se = 1 and the variances of gb and gq satisfy the 
  ## target microbiability and "heritability" 
  return(list('alpha' = beta_qtl / c(se), 
              'omega' = beta_otu / c(se),
              'qtl_list' = qtl_list,
              'se' = se / se ## = 1
  ))
}

#' Compute phenotype values based on generated object of current generation
#'
#' @inheritParams holo_simu
compute_phenotypes <- function(X,     #Genotypes matrix with 0,1,2 encoding
                               B,     #Microbiome matrix, CLR abundances
                               otu_list,
                               qtl_list, 
                               beta_qtl,
                               beta_otu,
                               Nqtl_y,
                               Notu_y,
                               se){
  
  # SNP part
  if (Nqtl_y>0) { #arrow from G to y
    #print(beta_qtl %>% head())
    X <- as.matrix(X)
    gq = t(X[qtl_list,]) %*% beta_qtl
  }
  
  # OTU part  
  if(Notu_y>0){ #arrow from B to y
    #print(beta_otu %>% head())
    B <- as.matrix(B)
    Botu = B[otu_list,]
    gb = t(Botu) %*% beta_otu
  }
  
  y = gq + gb + rnorm(length(gq), 0, se)
  return(list('y'=y, 'gq'=gq, 'gb'=gb))
  
}

#' apply mean on rows for each taxa across population
#' #'
#' @inheritParams holo_simu
compute_mean_microbiome <- function(microbiome, dir = F, n_ind = NULL, ao, mix.params){
  mean_microbiome <- rowMeans(microbiome)
  if(dir){
    stopif(is.null(n_ind))
    dir_mean <- rdirichlet(n_ind,as.numeric(mean_microbiome)*ao) %>% t()
    mix_mean <- (mix.params[1] * dir_mean + mix.params[2] * matrix(mean_microbiome,nrow=nrow(dir_mean),ncol=ncol(dir_mean),byrow=F))
    attr(mix_mean, "dirichlet") <- dir_mean
    return(mix_mean)
  }else{
    mean_microbiome
  }
}

#' Compute microbiomes for all individuals of current generation gathering all relevant object already computed.
#' #'
#' @inheritParams holo_simu
compute_current_microbiome <- function(beta,
                                       current_genotypes,
                                       mother_microbiomes,
                                       mean_microbiome,
                                       noise=0.1,
                                       taxa_scale, 
                                       lambda=0.5,
                                       dir = F,
                                       thetaX){
  
  beta_g_eps <- compute_beta_g(beta = beta, genotypes = current_genotypes, noise = noise, taxa_scale = taxa_scale)
  if(dir){
    internal_compute <- lambda*(mother_microbiomes) + (1-lambda)*mean_microbiome
  }else{
    internal_compute <- lambda*(mother_microbiomes) + (1-lambda)*matrix(mean_microbiome,nrow=nrow(beta_g_eps),ncol=ncol(beta_g_eps),byrow=F)
  }
  #TODO : check dimensions of thetaX
  if(!is.null(thetaX)){
    microbiome <- t(compositions::clr(t(internal_compute)%>% replace_zero())) + beta_g_eps + thetaX
  }else{
    microbiome <- t(compositions::clr(t(internal_compute)%>% replace_zero())) + beta_g_eps
  }
  colnames(microbiome) <- colnames(beta_g_eps)
  
  attr(microbiome,"noise_microbiome") <- list('attr'=attr(beta_g_eps,"noise"), 
                                              'clr_mic' = t(compositions::clr(t(internal_compute)%>% replace_zero())),
                                              'env_eff' = thetaX)
  #return CLR(microbiome)
  return(microbiome)
}

#' Replace zero in the matrix by 1e-15 to avoid infinite values when CLR transformation is computed
#'
#' @inheritParams holo_simu
replace_zero <- function(x){
  W <- oneOrDataset(x)
  nmv <- is.NMV(W)
  ifelse(nmv, W, 1e-15)
}

#' Obtain id of the parents selected for the next generation based on the criteria chosen by the user
#'
#' @inheritParams holo_simu
select_individual <- function(phenotypes,
                              microbiomes,
                              genotypes,
                              beta,
                              beta_otu,
                              selection, 
                              size_selection_F, 
                              size_selection_M, 
                              selection_type,
                              w.param){
  #extract id of individuals of interest
  n_F <- (size_selection_F * length(grep("^F",colnames(genotypes)))) |> round()
  n_M <- (size_selection_M * length(grep("^M",colnames(genotypes)))) |> round()
  
  #being able to select between GB, B, G, diversity or diversity + GB
  if(selection){
    ######
    #Few tests on parameters values
    #####
    if(is.null(size_selection_F) || is.null(size_selection_M)){
      print("Please check you provided size_selection_F and size_selection_M argument (int) for selection step")
    }
   if(!selection_type %in% c("GB", "B", "G", "diversity", "div.GB")){
     print("Selection type must be one of 'GB', 'B', 'G', 'diversity' or 'div.GB'")
   }
    if(selection_type == "GB"){
      score <- phenotypes$gq + as.vector(beta_otu %*% (beta[rowSums(beta) != 0, ] %*% genotypes))
      #score <- phenotypes$gq + as.vector((beta[rowSums(beta) != 0, ] %*% genotypes))
      #score <- phenotypes$gq + phenotypes$gb
    }else if(selection_type == "G"){
      score <- phenotypes$gq
    }else if(selection_type == "B"){
      score <- phenotypes$gb
    }else if(selection_type == "diversity"){
      score <- microbiomes %>% richness_from_abundances_gen() %>% select(Shannon) %>% as.matrix()
    }else{
      diversity <- microbiomes %>% richness_from_abundances_gen() %>% select(Shannon) %>% as.matrix()
      TBV <- phenotypes$gq + as.vector(beta_otu %*% (beta[rowSums(beta) != 0, ] %*% genotypes))
      score = w.param[1] * diversity + w.param[2] * TBV
    }
    
    #extract individuals ID
    F_id <- score[grep("^F",rownames(score)),] %>% sort(decreasing = T) %>% names() %>% str_extract("[0-9]+") %>% head(n_F)
    M_id <- score[grep("^M",rownames(score)),] %>% sort(decreasing = T) %>% names() %>% str_extract("[0-9]+") %>% head(n_M)
    
  }else{
    if(!is.null(size_selection_F) || !is.null(size_selection_M)){
      print("Warning : Selection is set to False but size_selection_F or size_selection_M is not NULL")
    }
    
    n_F <- (0.3 * length(grep("^F",colnames(genotypes)))) |> round()
    n_M <- (0.3 * length(grep("^M",colnames(genotypes)))) |> round()
    #extract individuals ID
    F_id <- colnames(genotypes)[grep("^F",colnames(genotypes))] %>% str_extract("[0-9]+") %>% sample(n_F)
    M_id <- colnames(genotypes)[grep("^M",colnames(genotypes))] %>% str_extract("[0-9]+") %>% sample(n_M)
  }

  return(list(F_id = F_id, M_id = M_id))
}
```

## Main function

```{r function-holo_simu}
#' Main function to generate transgenerational hologenomic data 
#'
#' @param h2 direct heritability value, between 0 and 1 
#' @param b2 microbiability value, between 0 and 1
#' @param founder_object output of generate_founder() function
#' @param otu_g percentage of taxa under genetic control, between 0 and 1, DEFAULT = 0.05
#' @param qtn_y number of causal SNPs for the phenotypes
#' @param n_ind number of individual per generation
#' @param n_gen number of generation, DEFAULT = 5
#' @param n_clust vector with taxa assignment, typical output of assign_taxa
#' @param selection bool, if selection process needed, DEFAULT = FALSE
#' @param size_selection_F percentage of female to select
#' @param size_selection_M percentage of male to select
#' @param selection_type mode of selection to be used, value in ("GB", "B", "G", "diversity", "div.GB"), DEFAULT = "GB"
#' @param seed seed value for samplings in the function
#' @param verbose bool, DEFAULT = T
#' @param lambda proportion of microbiome of offspring coming from vertical transmission, DEFAULT = 0.5
#' @param correlation Correlation between taxa within the same cluster, value between 0 and 1, DEFAULT = 0.5
#' @param mix.params Vector of two values giving the weights for the regularisation of the base population microbiome. mix.params[1] = weight for raw microbiome and mix.params[2] = weight for mean microbiome. DEFAULT = c(0.75,0.25)
#' @param mix.params.M
#' @param effect.size
#' @param noise.microbiome sd of microbiome noise, DEFAULT = 0.1
#' @param dir bool
#' @param ao for dirichlet sampling parameter
#' @param thetaX if environmental effect
#' @param env_gen vector of bool
#' @param w.param in case div.GB selection mode is chosen
#' 
#'
#' @return
#' A big list object with metada info such as beta matrix details and each generation at level 1. For each generation, the genotypes, the microbiomes, the phenotypes, the pedigree and the individuals selected can be reachable.
#' @rdname holo_simu
#' @export
holo_simu <- function(h2,
                      b2,
                      founder_object,
                      otu_g = 0.05,       
                      qtn_y = NULL,        
                      n_ind = NULL,       
                      n_gen = 5,         
                      n_clust = NULL,    
                      selection = F,       
                      size_selection_F = NULL,
                      size_selection_M = NULL,
                      selection_type = "GB",
                      seed=1234,
                      verbose = T,
                      lambda = 0.5,
                      correlation = 0.5,
                      mix.params = c(0.75,0.25),
                      mix.params.M = c(0.75,0.25),
                      effect.size = 0.1,
                      noise.microbiome = 0.1,
                      dir = F,
                      ao = 25,
                      thetaX = NULL,
                      env_gen = NULL,
                      w.param = c(0.5,0.5)){
  set.seed(seed)
  
  population <- attr(founder_object,"population")
  #obtain genotypes for all generation using MoBPS breeding.diploid
  if(is.null(n_ind)){
    n_ind <- ncol(founder_object)
  }
  
  ######
  #call to compute beta matrix function, common for all generations
  ######
  microbiome_filtered <- founder_object |> t()
  n_g <- nrow(get.geno(population, gen=1))
  
  #####################
  ### beta matrix generation
  #####################
    n_clust_g <- n_clust[n_clust != 0] %>% unique() %>% length()
    n_qtl_o <- ((n_g * 0.2)/n_clust_g)  %>% round()
    beta <- compute_beta_matrix_cluster(n_b = nrow(microbiome_filtered), 
                                        n_g = n_g, 
                                        n_clust = n_clust, 
                                        n_qtl_o = n_qtl_o, 
                                        n_otus = sum(n_clust != 0), 
                                        correlation = correlation, 
                                        effect_size = effect.size)
  beta_info <- attr(beta,"sim_params")
  otu_list <- list_c(beta_info$id_otu)
  
  ######
  #loop on generation to obtain microbiome
  ######
  ## ML estimate of founder pop compositions
  microbiome_gen <- apply(microbiome_filtered, 2, \(x) x/sum(x))
  ## shrinkage estimator: weighted average of ML composition and pop-level mean composition followed by clr transform
  microbiome_gen <- (mix.params[1] * microbiome_gen + mix.params[2] * matrix(rowMeans(microbiome_gen),nrow=nrow(microbiome_gen),ncol=ncol(microbiome_gen),byrow=F)) |> t() |> compositions::clr() |> t()
  # microbiome_gen <- microbiome_gen |> replace_zero() |> t() |> compositions::clr() |> t()
  #abundances
  taxa_scale <- microbiome_gen |> apply(1, sd)
  colnames(microbiome_gen) <- get.geno(population, gen = 1) %>% colnames()
  
  if(is.null(qtn_y)){
    Nqtl_y = 100
  }else{
    Nqtl_y = qtn_y
  }
  
  #####################
  ### founder phenotypes and parameters calibration
  #####################
  params_phenotypes <- calibrate_params_phenotypes(X0 = get.geno(population,gen = 1),
                                                   B0 = microbiome_gen, #CLR abundances ,
                                                   h2 = h2,
                                                   b2 = b2,
                                                   otu_list = otu_list,
                                                   Nqtl_y = Nqtl_y,
                                                   Notu_y = length(otu_list))
  
  phenotypes_founder <- compute_phenotypes(X = get.geno(population,gen = 1), 
                                           B = microbiome_gen, #CLR abundances 
                                           Nqtl_y = Nqtl_y, 
                                           Notu_y = length(otu_list), 
                                           otu_list = otu_list,
                                           qtl_list = params_phenotypes$qtl_list,
                                           beta_qtl = params_phenotypes$alpha,
                                           beta_otu = params_phenotypes$omega,
                                           se = params_phenotypes$se)
  
  
  #####################
  ### first elements output
  #####################
  metadata <- list(Beta_matrix = beta, 
                   QTN_y = params_phenotypes$qtl_list, 
                   G_coeff = params_phenotypes$alpha, 
                   B_coeff = params_phenotypes$omega)
  G0 <- list(microbiome = microbiome_gen |>  t() |> clrInv() |> t(), 
             genotypes =get.geno(population,gen = 1), 
             phenotypes = phenotypes_founder)
  list_output <- list(metadata = metadata, G0 = G0)
  
  
  #####################
  #Generation for loop
  ####################
  for(i in 1:n_gen){
    start_g <- proc.time()
    if(verbose){
      print(glue("Generation : {i}"))
    }
    #Warning : outside MoBPS functions, index of generation is i+1
    ######
    # Step 1 : get_current_genotype
    ######
    if(i==1){
      #renames microbiome matrix with individuals names
      population <- breeding.diploid(population,
                                     breeding.size = c(n_ind/2,n_ind/2),
                                     verbose = F)
    }else{
      population <- breeding.diploid(population,
                                     breeding.size = c(n_ind/2,n_ind/2), 
                                     selection.m.database = cbind(i,1,as.numeric(ID_selected$M_id)), 
                                     selection.f.database = cbind(i,2,as.numeric(ID_selected$F_id)),
                                     verbose = F)
    }
    
    ######
    # Step 2 : get_current_microbiome
    ######
    #Extract microbiomes of the mothers
    microbiome_gen <- microbiome_gen |>  t() |> clrInv() |> t()
    current_mother_microbiomes <- sapply(get.pedigree(population, gen=i+1)[,"mother"],function(x){
      microbiome_gen[,x]
    })
    
    mean_microbiome <- compute_mean_microbiome(microbiome_gen, 
                                               dir = dir,
                                               n_ind = n_ind,
                                               ao = ao,
                                               mix.params = mix.params.M)
    
    if(!is.null(env_gen)){
      if(env_gen[i]){
        microbiome_gen <- compute_current_microbiome(beta = beta,
                                                     current_genotypes = get.geno(population,gen = i+1),
                                                     mother_microbiomes = current_mother_microbiomes,
                                                     mean_microbiome = mean_microbiome,
                                                     noise = noise.microbiome, taxa_scale = taxa_scale, 
                                                     lambda = lambda,
                                                     dir = dir,
                                                     thetaX = thetaX) #not mandatory if we check dimension of mean_microbiome
      }else{
        microbiome_gen <- compute_current_microbiome(beta = beta,
                                                     current_genotypes = get.geno(population,gen = i+1),
                                                     mother_microbiomes = current_mother_microbiomes,
                                                     mean_microbiome = mean_microbiome,
                                                     noise = noise.microbiome, taxa_scale = taxa_scale, 
                                                     lambda = lambda,
                                                     dir = dir,
                                                     thetaX = thetaX)
      }
    }else{
      microbiome_gen <- compute_current_microbiome(beta = beta,
                                                   current_genotypes = get.geno(population,gen = i+1),
                                                   mother_microbiomes = current_mother_microbiomes,
                                                   mean_microbiome = mean_microbiome,
                                                   noise = noise.microbiome, taxa_scale = taxa_scale, 
                                                   lambda = lambda,
                                                   dir = dir,
                                                   thetaX = thetaX)
    }
    ######
    # Step 3 : get_current_phenotype
    ######
    phenotypes <- compute_phenotypes(X = get.geno(population,gen = i+1), 
                                     B = microbiome_gen, #CLR abundances 
                                     Nqtl_y = Nqtl_y, 
                                     Notu_y=length(otu_list), 
                                     otu_list = otu_list,
                                     qtl_list = params_phenotypes$qtl_list,
                                     beta_qtl = params_phenotypes$alpha,
                                     beta_otu = params_phenotypes$omega,
                                     se = params_phenotypes$se)
    
    
    ######
    # Step 4 : compute TBV to select individuals
    ######
    
    ID_selected <- select_individual(phenotypes = phenotypes,
                                     microbiomes = microbiome_gen |>  t() |> clrInv() |> t(),
                                     genotypes = get.geno(population,gen = 1+i),
                                     beta = beta,
                                     beta_otu = params_phenotypes$omega,
                                     selection = selection,
                                     size_selection_F = size_selection_F,
                                     size_selection_M = size_selection_M,
                                     selection_type = selection_type,
                                     w.param = w.param) 

    ###
    #update output
    ###
    list_output[[glue("G{i}")]] <- list(microbiome = microbiome_gen |>  t() |> clrInv() |> t(),
                                        genotypes = get.geno(population,gen = 1+i),
                                        phenotypes = phenotypes,
                                        pedigree = get.pedigree(population,gen = i+1),
                                        mean_microbiome = mean_microbiome,
                                        noise_microbiome = attr(microbiome_gen,"noise_microbiome"),
                                        ID_selected = c(glue("F_{ID_selected$F_id}"),glue("M_{ID_selected$M_id}")),
                                        omega_beta_g = as.vector(params_phenotypes$omega %*% (beta[rowSums(beta) != 0, ] %*% get.geno(population,gen = 1+i))))
    if(verbose){
      end_g <- proc.time() - start_g
      print(glue("Executed in {round(end_g[1] |> as.numeric(),3)} seconds"))
    }
  }
  return(list_output)
}

#' If selection based on diversity, need to go back to counts using multinomial sampling from abundances.
#'
#' @export
#' @inheritParams select_individual
richness_from_abundances_gen <- function(microbiome_matrix, n_loop=10, plot=T){
  microbiome_matrix[microbiome_matrix<0] <- 0
  for(i in 1:n_loop){
    microbiome_matrix <- apply(microbiome_matrix, 2, function(x){
      rmultinom(1,10000,as.vector(x))
    })
    tmp_physeq <- phyloseq(otu_table(microbiome_matrix,taxa_are_rows = T))
    if(i==1){
      richness <- estimate_richness(tmp_physeq,measures = c("Observed","Shannon","InvSimpson"))
    }else{
      richness <- richness + estimate_richness(tmp_physeq,measures = c("Observed","Shannon","InvSimpson"))
    }
  }
  return(richness/n_loop)
}

```

```{r examples-holo_simu}
```

```{r tests-holo_simu}
```

# Get functions 

# Plot functions 


<!--
# There can be development actions

Create a chunk with 'development' actions

- The chunk needs to be named `development` or `dev`
- It contains functions that are used for package development only
- Note that you may want to store most of these functions in the 0-dev_history.Rmd file

These are only included in the present flat template file, their content will not be part of the package anywhere else.
-->

```{r development-inflate, eval=FALSE}
# Keep eval=FALSE to avoid infinite loop in case you hit the knit button
# Execute in the console directly
fusen::inflate(flat_file = "dev/flat_first.Rmd", vignette_name = "Get started")
```


# Inflate your package

You're one inflate from paper to box.
Build your package from this very Rmd using `fusen::inflate()`

- Verify your `"DESCRIPTION"` file has been updated
- Verify your function is in `"R/"` directory
- Verify your test is in `"tests/testthat/"` directory
- Verify this Rmd appears in `"vignettes/"` directory
