<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Import Data â€¢ RITHMS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Import Data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">RITHMS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/generate-figures.html">Generate figures</a></li>
    <li><a class="dropdown-item" href="../articles/import-data.html">Data Importation</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/SolenePety/RITHMS"><span class="fa fab fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Import Data</h1>
            
      

      <div class="d-none name"><code>import-data.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RITHMS</span><span class="op">)</span></span></code></pre></div>
<!-- WARNING - This vignette is generated by {fusen} from dev/import-data.Rmd: do not edit by hand -->
<p>This vignette presents the steps required to import your own data
into the package. This mainly involves creating the
<code>founder-object</code> combining genetic and microbiotic
information, thanks to the <a href="https://github.com/tpook92/MoBPS" class="external-link"><code>MoBPS</code></a>
package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glue.tidyverse.org/" class="external-link">glue</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">MoBPS</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="about-accepted-formats">About accepted formats<a class="anchor" aria-label="anchor" href="#about-accepted-formats"></a>
</h2>
<p>The <code>MoBPS</code> package natively accepts two file formats to
generate the starting population with the
<code><a href="https://rdrr.io/pkg/MoBPS/man/creating.diploid.html" class="external-link">creating.diploid()</a></code> function: PED/MAP format and VCF. Thus,
it is possible to start a simulation from this kind of format, just by
specify the type with the <code>file_type</code> argument in the
<code><a href="../reference/read_input_data.html">read_input_data()</a></code> function.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create founder object from PED/MAP set</span></span>
<span><span class="va">founder_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_input_data.html">read_input_data</a></span><span class="op">(</span><span class="st">"path/to/microbiome.txt"</span>, <span class="st">"path/to/pedmap/'prefix'"</span>, biome_id_column <span class="op">=</span> <span class="st">"sample_id"</span>, file_type <span class="op">=</span> <span class="st">"pedmap"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Create founder object from VCF</span></span>
<span><span class="va">founder_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_input_data.html">read_input_data</a></span><span class="op">(</span><span class="st">"path/to/microbiome.txt"</span>, <span class="st">"path/to/vcf/'prefix'"</span>, biome_id_column <span class="op">=</span> <span class="st">"sample_id"</span>, file_type <span class="op">=</span> <span class="st">"vcf"</span><span class="op">)</span></span></code></pre></div>
<p>Microbiotic data is expected in the form of count table, with OTUs in
columns and individuals in rows. (A column dedicated to individual
identifiers is required). This data is filtered according to a
prevalence threshold thanks to the <code>threshold</code> argument in
the <code><a href="../reference/read_input_data.html">read_input_data()</a></code>.</p>
</div>
<div class="section level2">
<h2 id="import-simple-genotype-tables">Import simple genotype tables<a class="anchor" aria-label="anchor" href="#import-simple-genotype-tables"></a>
</h2>
<p>It is possible to provide RITHMS with a simple genotype matrix
encoded in the 0,1,2 format, with a prior file transformation, which we
detail below. The VCF file can be written by specifying the
<code>output_type</code> and <code>output_path</code> arguments.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_ind</span> <span class="op">&lt;-</span> <span class="fl">50</span></span>
<span><span class="va">n_snp</span> <span class="op">&lt;-</span> <span class="fl">50</span></span>
<span></span>
<span><span class="co"># Create a little genotype matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">geno_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">2</span>, size <span class="op">=</span> <span class="va">n_ind</span><span class="op">*</span><span class="va">n_snp</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                      nrow <span class="op">=</span> <span class="va">n_snp</span>, ncol <span class="op">=</span> <span class="va">n_ind</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Transform genotype matrix into VCF</span></span>
<span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transform_geno_into_vcf.html">transform_geno_into_vcf</a></span><span class="op">(</span><span class="va">geno_matrix</span>, output_type <span class="op">=</span> <span class="st">"dataframe"</span>, output_path <span class="op">=</span> <span class="st">"./output/path/file.vcf"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Then, you can continue with the classical import.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Create founder object from VCF</span></span>
<span><span class="va">founder_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_input_data.html">read_input_data</a></span><span class="op">(</span><span class="st">"path/to/microbiome.txt"</span>, <span class="st">"path/to/vcf/'prefix'"</span>, biome_id_column <span class="op">=</span> <span class="st">"sample_id"</span>, file_type <span class="op">=</span> <span class="st">"vcf"</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by SolÃ¨ne Pety, Ingrid David, Mahendra Mariadassou, Andrea Rau.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
