# WARNING - Generated by {fusen} from dev/flat_first.Rmd: do not edit by hand # nolint: line_length_linter.


#' Compute phenotype values based on generated objects of the current generation
#'
#' @param X A `matrix` of the current generation genotypes, encoding with 0, 1 and 2.
#' @param B A `matrix` of the CLR-transformed abundances of the current generation microbiome.
#' @param otu_list List of causal OTUs for the phenotypes.
#' @param qtl_list List of causal SNPs for the phenotypes.
#' @param beta_qtl Alpha, the regression coefficients corresponding to the QTL effects on the phenotype. See [calibrate_params_phenotypes()].
#' @param beta_otu Omega, the regression coefficients corresponding to taxa effects on the phenotype. See [calibrate_params_phenotypes()].
#' @param Nqtl_y Integer, number of causal SNPs for the phenotypes.
#' @param Notu_y Integer, number of causal OTUs for the phenotypes.
#' @param se Phenotypes Standard deviation, see [calibrate_params_phenotypes()].
#' 
#' @seealso [calibrate_params_phenotypes()]
#' @rdname compute_phenotypes

compute_phenotypes <- function(X,     #Genotypes matrix with 0,1,2 encoding
                               B,     #Microbiome matrix, CLR abundances
                               otu_list,
                               qtl_list, 
                               beta_qtl,
                               beta_otu,
                               Nqtl_y,
                               Notu_y,
                               se){
  
  # SNP part
  if (Nqtl_y>0) { #arrow from G to y
    #print(beta_qtl %>% head())
    X <- as.matrix(X)
    gq = t(X[qtl_list,]) %*% beta_qtl
  }
  
  # OTU part  
  if(Notu_y>0){ #arrow from B to y
    #print(beta_otu %>% head())
    B <- as.matrix(B)
    Botu = B[otu_list,]
    gb = t(Botu) %*% beta_otu
  }
  
  y = gq + gb + rnorm(length(gq), 0, se)
  return(list('y'=y, 'gq'=gq, 'gb'=gb))
  
}
