# WARNING - Generated by {fusen} from dev/flat_first.Rmd: do not edit by hand

#' Assign all taxa to a cluster, eventually under genetic control using hclust
#'
#' @param founder_object Output of read_input_data function
#' @param taxa_g Percentage of taxa under genetic control, DEFAULT = 0.1
#' @inheritParams stats::hclust
#' @inheritParams stats::cutree
#' @inheritParams vegan::vegdist
#'
#' @return
#' A vector with a length matching the total number of taxa with values from 0 to the number of taxa, 0 corresponding to the non under genetic control cluster
#' @rdname formatting_data
#' @export
#' @examples
#' #exemples with dataset

assign_taxa <- function(founder_object,
                        taxa_g = 0.1){
  dist_mat <- vegan::vegdist(x = t(founder_object),
               method = "bray")
  hc <- hclust(dist_mat)
  taxa_assign <- cutree(hc, k = 100)
  ntaxa_g = 0
  cluster_id_g = c()
  if(sum(table(taxa_assign)[table(taxa_assign) > 10 & table(taxa_assign) < 25]) < round(length(taxa_assign)*taxa_g)){
    print(glue("{sum(table(taxa_assign)[table(taxa_assign) > 10 & table(taxa_assign) < 25])} taxa available for genetic control sampling but {round(length(taxa_assign)*taxa_g)} asked. Change taxa_g parameter please."))
  }
  while(ntaxa_g < round(length(taxa_assign)*taxa_g)){
    new_id <- table(taxa_assign)[table(taxa_assign) > 10 & table(taxa_assign) < 25] %>% sample(1) %>% names() %>% as.numeric()
    if(! new_id %in% cluster_id_g){
      cluster_id_g <- append(cluster_id_g, new_id)
      ntaxa_g <- ntaxa_g + sum(taxa_assign == new_id)
    }
  }
  taxa_assign_g <- ifelse(taxa_assign %in% cluster_id_g, taxa_assign, 0) %>% as.factor()
  levels(taxa_assign_g) <- c(0:length(cluster_id_g))
  return(taxa_assign_g)
}


#' Formatting data from file paths to base population object
#'
#' @param path_to_microbiome String giving the path to count table file
#' @param path_to_pedmap String giving the path and prefix to ped and map file
#' @param threshold Threshold for rarefaction, DEFAULT = 0.05
#' @param ind_selected Vector of string values with individuals to keep, have to match rownames of count table file, DEFAULT = NULL
#'
#' @return
#' A data.frame corresponding to (rarefied) microbiome with individuals in rows and taxa in columns. Genotypes data.frame is an attribute called "population" reachable using attr(output_name,"population")
#' @rdname formatting_data
#' @export
read_input_data <- function(path_to_microbiome,
                            path_to_pedmap,
                            threshold = 0.05,
                            ind_selected = NULL){
  ########
  #read microbiome file and make genotypes and microbiome data match according to individuals
  ########
  microbiome <- fread(path_to_microbiome)
  microbiome_filtered5 <- rarefied_microbiome(microbiome, threshold = threshold)
  if(!is.null(ind_selected)){
    microbiome_filtered5 <- microbiome_filtered5 %>% filter(rownames(.) %in% ind_selected)
  }
  founder_object <- generate_founder(path = glue("{path_to_pedmap}"),
                                     microbiome_filtered5)
  return(founder_object)
}

#' Rarefaction step within read_input_data
#'
#' @inheritParams read_input_data
rarefied_microbiome <- function(microbiome_matrix, 
                                threshold = 0.05){
  microbiome_matrix <- microbiome_matrix |> as_tibble() |> tibble::column_to_rownames("numero_individu")
  abundant_otus <- tibble(
    V1         = colnames(microbiome_matrix),
    prevalence = colMeans(microbiome_matrix > 0)
  ) |> rename(OTU = V1) |> filter(prevalence >= threshold) |> pull(OTU)
  return(microbiome_matrix[, abundant_otus])
}

#' Formatting of ped and map files into haplotypes table and filter individuals
#'
#' @inheritParams read_input_data
generate_founder <- function(path = NULL,
                             microbiote_matrix){
  # # Convert PED-file into haplotype dataset (one haplotype per colum)
  if(is.null(path)){
    stop("path must be a valid string like : '/path/to/directory/prefix'")
  }else{
    if(FALSE %in% file_test("-f", c(glue::glue("{path}.ped"),
                                    glue::glue("{path}.map")))){
      stop(glue::glue("Files {path}.ped and/or {path}.map don't exist, please check your path argument"))
    }else{
      map <- as.matrix(read.table(glue::glue("{path}.map")))
      ped <- as.matrix(read.table(glue::glue("{path}.ped")))
      #Step to filter ped map based on microbiome matrix,
      #rownames microbiote_matrix must be ind ID to match column 2 of ped file
      ped_ind <- ped[,2] %>% gsub("\\s","",.)
      microbiote_ind <- rownames(microbiote_matrix)
      microbiote_filtered <- microbiote_matrix[microbiote_ind %in% ped_ind,]
      ped <- ped[ped_ind %in% microbiote_ind,]
      ####
      nsnp <- (ncol(ped)-6)/2
      haplo1 <- ped[,1:nsnp*2+6-1] #all odd numbers
      haplo2 <- ped[,1:nsnp*2+6] #all uneven numbers
      haplo <- t(rbind(haplo1, haplo2)[c(0,nrow(haplo1)) + sort(rep(1:nrow(haplo1),2)),])
      #haplo <- ifelse(haplo == "0", NA,haplo)
      population <- creating.diploid(dataset = haplo, map = map, verbose=verbose)
      attr(microbiote_filtered,"population") <- population
      return(microbiote_filtered)
    }
  }
}

