# WARNING - Generated by {fusen} from dev/flat_first.Rmd: do not edit by hand # nolint: line_length_linter.

#' Formatting data from file paths to base population object
#'
#' @importFrom data.table fread
#' @importFrom dplyr filter
#' @importFrom glue glue
#' @importFrom magrittr %>%
#'
#' @param path_to_microbiome String giving the path to count table file. `"path/to/microbiome.{extension}"`
#' @param path_to_genotype String giving the path and prefix to the genotype file. This should be one of `"pedmap"` or `"vcf"`. (default: `"pedmap"`), see `file_type` param. For a `"example_pedmap.ped"` file enter only `"path/to/example_pedmap"`.
#' @param file_type String specifying the file type used to load genotype data. This should be one of `"pedmap"` or `"vcf"`. (default: `"pedmap"`).
#' @param biome_id_column String specifying the name of the column containing the individual IDs for the microbiome matrix. Default is "ind_id".
#' @param threshold Threshold for rarefaction, DEFAULT = 0.05
#' @param ind_selected Vector of string values with individuals to keep, have to match rownames of count table file, DEFAULT = NULL
#' 
#' @return
#' A `list` containing two object.
#' A `matrix` containing the microbiome(`microbiome`). 
#' A `list` from [MoBPS](https://pubmed.ncbi.nlm.nih.gov/32229505/) package, containing the genotypes (`population`). Individuals are in columns and taxa or SNPs are in rows.
#' 
#' @note
#' The number of individuals in the microbiome an genotype data must be equivalent.
#' 
#' @seealso [generate_founder()], [rarefied_microbiome()]
#' 
#' @rdname read_input_data
#' @export
#' @examples
#' \dontrun{
#' # Create founder object from PED/MAP set
#' founder_object <- read_input_data(path_to_microbiome = "/path/to/microbiome.txt",
#'                                   path_to_pedmap = "/path/to/pedmap/'prefix'",
#'                                   biome_id_column = "ind_id",
#'                                   threshold = 0.05,
#'                                   ind_selected = NULL,
#'                                   file_type = "pedmap")
#'
#' # Create founder object from VCF
#' founder_object <- read_input_data(path_to_microbiome = "/path/to/microbiome.txt",
#'                                   path_to_pedmap = "/path/to/vcf/'prefix'",
#'                                   biome_id_column = "ind_id",
#'                                   threshold = 0.05,
#'                                   ind_selected = NULL,
#'                                   file_type = "vcf")
#' }
read_input_data <- function(path_to_microbiome,
                            path_to_genotype,
                            file_type = "pedmap",
                            biome_id_column = "ind_id",
                            threshold = 0.05,
                            ind_selected = NULL){
  ########
  #read microbiome file and make genotypes and microbiome data match according to individuals
  ########
  microbiome <- data.table::fread(path_to_microbiome)
  microbiome_filtered5 <- rarefied_microbiome(microbiome, id_column = biome_id_column, threshold = threshold)
  if(!is.null(ind_selected)){
    microbiome_filtered5 <- microbiome_filtered5 %>% filter(rownames(.) %in% ind_selected)
  }
  
  if(!file_type %in% c("pedmap", "vcf")){
    print("The genotype file type must be one of 'pedmap' or 'vcf'.")
  } else {
    founder_object <- generate_founder(path = glue::glue("{path_to_genotype}"), microbiome_filtered5, file_type = file_type)

  }
  
  return(founder_object)
}

