# WARNING - Generated by {fusen} from dev/flat_first.Rmd: do not edit by hand # nolint: line_length_linter.

#' Convert a 0/1/2 genotype matrix into a VCF-like format
#'
#' This function converts a genotype matrix encoded as 0,1,2 into a VCF-like format. It can either return the VCF content as a `data.frame`, write it to a .vcf file or do both.
#'
#' @param geno_matrix A matrix of genotypes with values 0, 1, 2. SNPs are in rows and individuals in columns.
#' @param output_type A character, that specifies the output type. Choose between `"file"` (write `.vcf`),
#'                    `"dataframe"` (return `data.frame`), or `"both"` (do both).
#' @param output_path A character string that specifies the output path. Requiered if `output_type` is "file" or "both".
#' 
#' @return A `data.frame` if `output_type = "dataframe" or "both"`, or just write the `.vcf` file if `output_type = "file" or "both"`.
#' 
#' @rdname transform_geno_into_vcf
#' @export
#' @examples
#' n_ind <- 50
#' n_snp <- 50
#'
#' # Create a little genotype matrix
#' set.seed(123)
#' geno_matrix <- matrix(sample(0:2, size = n_ind*n_snp, replace = TRUE),
#'                       nrow = n_snp, ncol = n_ind)
#'
#' # Transform genotype matrix into VCF
#' vcf <- transform_geno_into_vcf(geno_matrix, output_type = "dataframe")
#' head(vcf[1:3,1:7])

transform_geno_into_vcf <- function(geno_matrix, output_type = c("file", "dataframe", "both"), output_path = NULL){
  
    n_ind = ncol(geno_matrix)
    n_snp = nrow(geno_matrix)
    
    convert_to_haplo <- function(x) {
    if(is.na(x)) return(".")
    if (x == 0) return("0/0")
    if (x == 1) return("0/1")
    if (x == 2) return("1/1")
    }
  
    geno_matrix_haplo <- apply(geno_matrix, c(1,2), convert_to_haplo)
  
    vcf_header <- c(
      "##fileformat=VCFv4.2",
      paste0("##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">"),
      paste("#CHROM", "POS", "ID", "REF", "ALT", "QUAL", "FILTER", "INFO", "FORMAT", paste0("Indiv_", 1:n_ind), sep="\t")
    )
  
    vcf_snps <- data.frame(
      CHROM = "1", 
      POS = 1:n_snp, 
      ID = paste0("rs", 1:n_snp), 
      REF = "A", ALT = "T", 
      QUAL = ".", FILTER = ".", INFO = ".", FORMAT = "GT",
      geno_matrix_haplo
    )
    
    if(output_type %in% c("file", "both")){
      write.table(vcf_snps, file = output_path, sep = "\t", quote = FALSE, row.names = TRUE, col.names = TRUE)
      
    } else if(output_type %in% c("dataframe", "both")){
      return(vcf_snps)
    } else{
      print("The output type must be one of 'file' or 'dataframe'.")
    }
}
